runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/internal/registry/FileWatcherTask.java

```diff
@@ -0,0 +1,147 @@
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+
+public class FileWatcherTask extends WatcherTask
+{
+    private final Map<Path, byte[]> configHashes;
+    private WatchService watchService;
+
+    public FileWatcherTask(BiConsumer<URL, String> configChangeListener)
+    {
+        super(configChangeListener);
+        this.configHashes = new HashMap<>();
+        try
+        {
+            watchService = FileSystems.getDefault().newWatchService();
+        }
+        catch (IOException ex)
+        {
+            rethrowUnchecked(ex);
+        }
+
+    }
+
+    @Override
+    public boolean run()
+    {
+        try
+        {
+            while (true)
+            {
+                try
+                {
+                    final WatchKey key = watchService.take();
+
+                    for (WatchEvent<?> event : key.pollEvents())
+                    {
+                        final Path changed = ((Path) key.watchable()).resolve((Path) event.context());
+                        if (configHashes.containsKey(changed))
+                        {
+                            String configText = readConfigText(configURLs.get(changed));
+                            byte[] newConfigHash = computeHash(configText);
+                            if (!Arrays.equals(configHashes.get(changed), newConfigHash))
+                            {
+                                configHashes.put(changed, newConfigHash);
+                                configChangeListener.accept(configURLs.get(changed), configText);
+                            }
+                        }
+                    }
+                    key.reset();
+                }
+                catch (InterruptedException ex)
+                {
+                    watchService.close();
+                    break;
+                }
+            }
+        }
+        catch (IOException ex)
+        {
+            rethrowUnchecked(ex);
+        }
+        return true;
+    }
+
+    @Override
+    protected void doInitialConfiguration(URL configURL)
+    {
+        Path configPath = Paths.get(new File(configURL.getPath()).getAbsolutePath());
+        configURLs.put(configPath, configURL);
+        String configText = readConfigText(configURL);
+        configHashes.put(configPath, computeHash(configText));
+
+        try
+        {
+            configPath.getParent().register(watchService, ENTRY_MODIFY, ENTRY_CREATE, ENTRY_DELETE);
+        }
+        catch (IOException ignored)
+        {
+        }
+
+        configChangeListener.accept(configURL, configText);
+    }
+
+    private String readConfigText(URL configURL)
```

Comment:
Simpler to pass `Path` here instead of `URL`, as we are dealing only with `file` scheme URLs.

Reply:
It's not enoguh, since we have configfiles inside jars. If we're using only Paths, that will be a FileSystemNotFoundException.